File: ocaml/lib3d.ml
type vec3 = {x : float; y : float; z : float};;
type mat3 = {
  m11 : float; m12 : float; m13 : float;
  m21 : float; m22 : float; m23 : float;
  m31 : float; m32 : float; m33 : float;
};;

let negate v = {
  x = -.v.x;
  y = -.v.y;
  z = -.v.z;
};;

let add a b = {
  x = a.x +. b.x;
  y = a.y +. b.y;
  z = a.z +. b.z;
};;

let subtract a b = {
  x = a.x -. b.x;
  y = a.y -. b.y;
  z = a.z -. b.z;
};;

let mul_vec_sf v sf = {
  x = v.x *. sf;
  y = v.y *. sf;
  z = v.z *. sf;
};;

let mul_vec_mat v m = {
  x = v.x *. m.m11 +. v.y *. m.m21 +. v.z *. m.m31;
  y = v.x *. m.m12 +. v.y *. m.m22 +. v.z *. m.m32;
  z = v.x *. m.m13 +. v.y *. m.m23 +. v.z *. m.m33;
};;

let mul_mat_mat a b =
  let p = mul_vec_mat {x = a.m11; y = a.m12; z = a.m13} b in
  let q = mul_vec_mat {x = a.m21; y = a.m22; z = a.m23} b in
  let r = mul_vec_mat {x = a.m31; y = a.m32; z = a.m33} b in
  {
    m11 = p.x; m12 = p.y; m13 = p.z;
    m21 = q.x; m22 = q.y; m23 = q.z;
    m31 = r.x; m32 = r.y; m33 = r.z;
  };;

let divide v = function
  | 0. -> {x = 0.; y = 0.; z = 0.}
  | k  -> {
      x = v.x /. k;
      y = v.y /. k;
      z = v.z /. k;
    };;

let dotProduct a b =
  a.x *. b.x +.
  a.y *. b.y +.
  a.z *. b.z;;

let crossProduct a b = {
  x = a.y *. b.z -. a.z *. b.y;
  y = a.z *. b.x -. a.x *. b.z;
  z = a.x *. b.y -. a.y *. b.x;
};;

let tripleProduct a b c = dotProduct a (crossProduct b c);;

let magnitude v =
  sqrt (
    v.x *. v.x +.
    v.y *. v.y +.
    v.z *. v.z);;

let distance a b = magnitude (subtract a b);;

let normalize v = divide v (magnitude v);;

let xAxisRotationMatrix theta =
  let st = sin theta in
  let ct = cos theta in
  {
    m11 = 1.; m12 = 0.; m13 = 0.;
    m21 = 0.; m22 = ct; m23 = st;
    m31 = 0.; m32 = st; m33 = ct;
  };;

let yAxisRotationMatrix theta =
  let st = sin theta in
  let ct = cos theta in
  {
    m11 = ct; m12 = 0.; m13 = st;
    m21 = 0.; m22 = 1.; m23 = 0.;
    m31 = st; m32 = 0.; m33 = ct;
  };;

let zAxisRotationMatrix theta =
  let st = sin theta in
  let ct = cos theta in
  {
    m11 =   ct; m12 = st; m13 = 0.;
    m21 = -.st; m22 = ct; m23 = 0.;
    m31 =   0.; m32 = 0.; m33 = 1.;
  };;

let axisRotationMatrix v theta =
  let v = normalize v in
  let st = sin theta in
  let ct = cos theta in
  let nc = 1. -. ct in
  {
    m11 = v.x *. v.x *. nc +. ct;
    m12 = v.x *. v.y *. nc +. v.z *. st;
    m13 = v.x *. v.z *. nc -. v.y *. st;
    m21 = v.x *. v.y *. nc -. v.x *. st;
    m22 = v.y *. v.y *. nc +. ct;
    m23 = v.y *. v.z *. nc +. v.x *. st;
    m31 = v.x *. v.z *. nc +. v.y *. st;
    m32 = v.y *. v.z *. nc -. v.x *. st;
    m33 = v.z *. v.z *. nc +. ct;
  };;

let scale_sf sf = {
  m11 = sf; m12 = 0.; m13 = 0.;
  m21 = 0.; m22 = sf; m23 = 0.;
  m31 = 0.; m32 = 0.; m33 = sf;
};;

let scale_xyz sx sy sz = {
  m11 = sx; m12 = 0.; m13 = 0.;
  m21 = 0.; m22 = sy; m23 = 0.;
  m31 = 0.; m32 = 0.; m33 = sz;
};;

let scale_axis v sf =
  let v = normalize v in
  let sf = sf -. 1. in
  {
    m11 = sf *. v.x *. v.x +. 1.;
    m12 = sf *. v.x *. v.y;
    m13 = sf *. v.x *. v.z;
    m21 = sf *. v.x *. v.y;
    m22 = sf *. v.y *. v.y +. 1.;
    m23 = sf *. v.y *. v.z;
    m31 = sf *. v.x *. v.z;
    m32 = sf *. v.y *. v.z;
    m33 = sf *. v.z *. v.z +. 1.;
  };;

let orthoProjection v = scale_axis v 0.;;

let reflection v = scale_axis v (-1.);;

File: c++/lib3d.cpp
#include <cmath>
#include "lib3d.h"

using namespace lib3d;

Vector3 lib3d::negate(const Vector3 v)
{
    Vector3 v2;
    v2.x = -v.x;
    v2.y = -v.y;
    v2.z = -v.z;
    return v2;
}

Vector3 lib3d::add(const Vector3 a, const Vector3 b)
{
    Vector3 v;
    v.x = a.x + b.x;
    v.y = a.y + b.y;
    v.z = a.z + b.z;
    return v;
}

Vector3 lib3d::subtract(const Vector3 a, const Vector3 b)
{
    Vector3 v;
    v.x = a.x - b.x;
    v.y = a.y - b.y;
    v.z = a.z - b.z;
    return v;
}

Vector3 lib3d::multiply(const Vector3 v, const double k)
{
    Vector3 v2;
    v2.x = v.x * k;
    v2.y = v.y * k;
    v2.z = v.z * k;
    return v2;
}

Vector3 lib3d::multiply(const Vector3 v, const Matrix3 M)
{
    Vector3 v2;
    v2.x = v.x * M.m11 + v.y * M.m21 + v.z * M.m31;
    v2.y = v.x * M.m12 + v.y * M.m22 + v.z * M.m32;
    v2.z = v.x * M.m13 + v.y * M.m23 + v.z * M.m33;
    return v2;
}

Matrix3 lib3d::multiply(const Matrix3 A, const Matrix3 B)
{
    /* Can be sped up */

    Vector3 p;
    p.x = A.m11;
    p.y = A.m12;
    p.z = A.m13;
    Vector3 q;
    q.x = A.m21;
    q.y = A.m22;
    q.z = A.m23;
    Vector3 r;
    r.x = A.m31;
    r.y = A.m32;
    r.z = A.m33;

    p = multiply(p, B);
    q = multiply(q, B);
    r = multiply(r, B);

    Matrix3 M;
    M.m11 = p.x;
    M.m12 = p.y;
    M.m13 = p.z;
    M.m21 = q.x;
    M.m22 = q.y;
    M.m23 = q.z;
    M.m31 = r.x;
    M.m32 = r.y;
    M.m33 = r.z;
    return M;
}

Vector3 lib3d::divide(const Vector3 v, const double k)
{
    if(k == 0)
    {
        return Vector3{0, 0, 0};
    }

    Vector3 v2;
    v2.x = v.x / k;
    v2.y = v.y / k;
    v2.z = v.z / k;
    return v2;
}

double lib3d::dotProduct(const Vector3 a, const Vector3 b)
{
    return a.x * b.x
         + a.y * b.y
         + a.z * b.z;
}

Vector3 lib3d::crossProduct(const Vector3 a, const Vector3 b)
{
    Vector3 v;
    v.x = a.y * b.z - a.z * b.y;
    v.y = a.z * b.x - a.x * b.z;
    v.z = a.x * b.y - a.y * b.x;
    return v;
}

double lib3d::tripleProduct(const Vector3 a, const Vector3 b, const Vector3 c)
{
    return dotProduct(a, crossProduct(b, c));
}

double lib3d::magnitude(const Vector3 v)
{
    return sqrt(v.x * v.x
              + v.y * v.y
              + v.z * v.z);
}

double lib3d::distance(const Vector3 a, const Vector3 b)
{
    return magnitude(subtract(a, b));
}

Vector3 lib3d::normalize(const Vector3 v)
{
    return divide(v, magnitude(v));
}

Matrix3 lib3d::xAxisRotationMatrix(const double theta)
{
    double sinTheta = sin(theta);
    double cosTheta = cos(theta);

    Matrix3 M;
    M.m11 = 1;
    M.m12 = 0;
    M.m13 = 0;
    M.m21 = 0;
    M.m22 = cosTheta;
    M.m23 = sinTheta;
    M.m31 = 0;
    M.m32 = sinTheta;
    M.m33 = cosTheta;
    return M;
}

Matrix3 lib3d::yAxisRotationMatrix(const double theta)
{
    double sinTheta = sin(theta);
    double cosTheta = cos(theta);

    Matrix3 M;
    M.m11 = cosTheta;
    M.m12 = 0;
    M.m13 = sinTheta;
    M.m21 = 0;
    M.m22 = 1;
    M.m23 = 0;
    M.m31 = sinTheta;
    M.m32 = 0;
    M.m33 = cosTheta;
    return M;
}

Matrix3 lib3d::zAxisRotationMatrix(const double theta)
{
    double sinTheta = sin(theta);
    double cosTheta = cos(theta);

    Matrix3 M;
    M.m11 = cosTheta;
    M.m12 = sinTheta;
    M.m13 = 0;
    M.m21 = -sinTheta;
    M.m22 = cosTheta;
    M.m23 = 0;
    M.m31 = 0;
    M.m32 = 0;
    M.m33 = 1;
    return M;
}

Matrix3 lib3d::axisRotationMatrix(const Vector3 v, const double theta)
{
    /* Just in case */
    v = normalize(v);

    double sinTheta = sin(theta);
    double cosTheta = cos(theta);
    double negCos = 1 - cosTheta;

    Matrix3 M;
    M.m11 = v.x * v.x * negCos + cosTheta;
    M.m12 = v.x * v.y * negCos + sinTheta * v.z;
    M.m13 = v.x * v.z * negCos - sinTheta * v.y;
    M.m21 = v.x * v.y * negCos - sinTheta * v.z;
    M.m22 = v.y * v.y * negCos + cosTheta;
    M.m23 = v.y * v.z * negCos + sinTheta * v.x;
    M.m31 = v.x * v.z * negCos + sinTheta * v.y;
    M.m32 = v.y * v.z * negCos - sinTheta * v.x;
    M.m33 = v.z * v.z * negCos + cosTheta;
    return M;
}

Matrix3 lib3d::scale(const double sf)
{
    Matrix3 M;
    M.m11 = M.m22 = M.m33 = sf;
    M.m12 = M.m13 = M.m21 = M.m23 = M.m31 = M.m32 = 0;
    return M;
}

Matrix3 lib3d::scale(const double sfx, const double sfy, const double sfz)
{
    Matrix3 M;
    M.m11 = sfx;
    M.m22 = sfy;
    M.m33 = sfz;
    M.m12 = M.m13 = M.m21 = M.m23 = M.m31 = M.m32 = 0;
    return M;
}

Matrix3 lib3d::scale(const Vector3 v, const double sf)
{
    /* Just in case */
    v = normalize(v);

    sf = sf - 1;

    Matrix3 M;
    M.m11 = sf * v.x * v.x + 1;
    M.m12 = sf * v.x * v.y;
    M.m13 = sf * v.x * v.z;
    M.m21 = sf * v.y * v.x;
    M.m22 = sf * v.y * v.y + 1;
    M.m23 = sf * v.y * v.z;
    M.m31 = sf * v.z * v.x;
    M.m32 = sf * v.z * v.y;
    M.m33 = sf * v.z * v.z + 1;
    return M;
}

Matrix3 lib3d::orthoProjection(const Vector3 v)
{
    /* Can be sped up */
    return scale(v, 0);
}

Matrix3 lib3d::reflection(const Vector3 v)
{
    /* Can be sped up */
    return scale(v, -1);
}
